firewood/src/proofs/diff.rs

```diff
@@ -0,0 +1,2028 @@
+            }
+        };
+        #[cfg(feature = "ethhash")]
+        let node = match node {
+            Child::Node(node) => node.clone().into(),
+            Child::AddressWithHash(addr, hash) => {
+                node_hash = Some(hash.into());
+                self.trie.read_node(addr)?
+            }
+            Child::MaybePersisted(maybe_persisted, hash) => {
+                node_hash = Some(hash.into());
+                maybe_persisted.as_shared_node(&self.trie)?
+            }
+        };
+
+        // Compute the full path for the node. This is used to compare with the full path
+        // of the current node from the other trie.
+        let full_path = Path::from_nibbles_iterator(
+            pre_path
+                .iter()
+                .copied()
+                .chain(node.partial_path().iter().copied()),
+        );
+        Ok(NodeState {
+            path: full_path,
+            node,
+            hash: node_hash,
+        })
+    }
+
+    fn next(&mut self) -> Result<Option<NodeState>, FileIoError> {
+        counter!("firewood.diff_merkle", "traversal" => "next").increment(1);
+        // Pop the next node state from the stack and reset `prev_num_children`
+        self.prev_num_children = 0;
+        if let Some(frame) = self.stack.pop() {
+            let state = self.create_node_state(&frame.pre_path, frame.node, frame.hash)?;
+            // If this node is a branch, push its children onto the stack so they will be processed next.
+            if let Node::Branch(branch) = &*state.node {
+                // Since a stack is LIFO and we want to perform pre-order traversal, we pushed the
+                // children in reverse order.
+                for (path_comp, child) in branch.children.clone().into_iter().rev() {
+                    if let Some(child) = child {
+                        // Generate the pre-path for this child, and push it onto the traversal stack.
+                        let child_pre_path = Path::from_nibbles_iterator(
+                            state.path.iter().copied().chain(once(path_comp.as_u8())),
+                        );
+                        self.stack.push(TraversalStackFrame {
+                            pre_path: child_pre_path,
+                            node: child,
+                            hash: None,
+                        });
+                        // Increment the number of children that has been added to the stack.
+                        self.prev_num_children += 1;
+                    }
+                }
+            }
+            // Return the state of the current node
+            Ok(Some(state))
+        } else {
+            // Stack is empty, iteration is complete. Return None.
+            Ok(None)
+        }
+    }
+
+    /// Used to not iterate through the children of a branch node if there is a hash match.
+    fn skip_children(&mut self) {
+        for _ in 0..self.prev_num_children {
+            self.stack.pop();
+        }
```

Comment:
Use `Vec::truncate` here which will be faster:

```suggestion
        let truncated_length = self.stack.len().saturating_sub(self.prev_num_children);
        self.stack.truncate(truncated_length);
```

This also calls out that you should check that there are N elements in the stack before doing this, perhaps as a `debug_assert`.